### ISSUE TITLE
Window.wait() timeout and use with window created after invocation doesn't seem to work

### ISSUE THREAD

**Issue by zxrobinson** [spyoungtech/ahk#131]
Hi,

I'm trying to wait for a window to open, but Window.wait() doesn't stop waiting when the window is created and it doesn't seem to use the timeout value.  Here's the code I'm using:

```
ahk = AHK()

win3 = ahk.win_get(title='Untitled - Notepad')
result = win3.exists()
print("result = %s" % (result))

print("Waiting started")
result = win3.wait(seconds_to_wait=10)
"""
At this point I open Notepad manually
"""
print("Waiting ended with result: %s" % (result))
```
The output I get is this:
```
result = False
Waiting started
```
And the process just sits there. I've tried using arguments in quotes like '10' as well as a regular integer, but there isn't any difference in the result. To terminate the python script, I have to kill the AutoHotKey process. When I do that, the output I get is:

`Waiting ended with result: None`

Am I doing something wrong, or is this a bug. I'm using Windows 10 64 bit, Python 3.8.10 64 bit, AutoHotKey 1.1.24.01 64 bit Unicode. I looked for tests for waiting for a non-existent window, but I didn't see any...

**Comment by spyoungtech** [comment]
Thank you for bringing this to my attention @zxrobinson 

I see at least two major issues in the library here. (1) the `wait` method behavior makes no sense as currently implemented and (2) `win_get` shouldn't be allowed to return a Window object if no such window exists.


# WinWait not properly implemented

This `WinWait` feature is not yet documented in the public API. I don't think it's currently tested or implemented properly. The current implementation will simply wait for the window of a matching hwnd, which doesn't make much sense to me. If the hwnd was obtained, then the window exists so there's no reason to wait on it... if a Window, say, by the same title, ever ceases to exist and comes back, it probably will have a new hwnd, which is probably not how users expect this to work.

The correct, or at least more intuitive implementation, in my view, would be that:

`Window.wait` (and `AsyncWindow.wait` -- or `ahk.win_wait` as an alias) should be a _classmethod_ for and it would return a new `Window` object with a valid hwnd if successful. In the case where the window is not found within the timeout, it should (by default) raise a `TimeoutError` -- or maybe return `None`.

# win_get should raise an error if the specified window does not exist.

Another peculiarity is that `win_get` seems to be allowed to return a `Window` object even when a window is not found. The hwnd will simply be an empty string. However, I don't think this ever makes any sense, as none of the commands will work properly in this case. In _any_ case where a window is not found, a `Window` object should probably not ever be returned (unless that object is redesigned significantly)... in this case where you call `win_get` and the window doesn't exist a `ahk.window.WindowNotFoundError` should be raised, like what happens when using the `ahk.find_window` methods.

**Comment by zxrobinson** [comment]
Would it make sense to put ahk.win_get and win.exist on the Window class, along with a class Window.wait()? It also make sense to create two APIs for Window.wait, one that throws an exception and a second one that wraps it, catches the exception and returns None.

What do you use to debug when working on this. I'm asking because I wonder about trying to do something about these issues.

**Comment by spyoungtech** [comment]
> might make sense to create two APIs for Window.wait, one that throws an exception and a second one that wraps it, catches the exception and returns None

Yeah, that's a good thought @zxrobinson --  I'm trying to think of similar situations in other popular libraries and how this is typically handled. While I'm not _too_ concerned about backwards compatibility since we've not yet published a v1.0 -- having two APIs would provide the ability to maintain the current API without a breaking change. 

I think raising `TimeoutError` is the correct default for `WinWait` when not found. Maybe a keyword argument could change the behavior. `.wait(10, suppress=True)` 🤷 


> What do you use to debug when working on this. 

Nothing special. I just use PyCharm as my editor and I usually test/debug by writing/running unit tests or use an interactive prompt (like iPython/Jupyter). To debug further, I might enable debug logging (`logging.basicConfig(level=logging.DEBUG)`) which will show the AHK scripts being executed for each method call, or I may prototype some code AHK and run it directly using autohotkey.exe or the `run_script` interface.

Hope that helps.

**Comment by zxrobinson** [comment]
I have a class method version of wait that seems to work, along with a new template for wait and a test class file. The test passes with a warning that a subprocess is still running in the test. Can you look at the classmethod to see if it does what ought to be done?

[ahk-wait-window.zip](https://github.com/spyoungtech/ahk/files/7645733/ahk-wait-window.zip)
e?
